### PR TITLE
[5.2] Container can get contextual dependency for method by looking at method's class object

### PR DESCRIPTION
**Pull request by hannesvdvreken**
Usage:

``` php
class Foo
{
    public function bar(QueueContract $queue)
    {
        ...
    }
}

$container->when(Foo::class)->needs(QueueContract::class)->give(Queue::class);

$container->call('Foo@bar');
```

Also more fine grained contextual binding is possible:

``` php
$container->when('Foo@bar')->needs(QueueContract::class)->give(Queue::class);
```

This will get priority over any `->when(Foo::class)` contextual binding.

Works in controller methods as well.

(Follow up PR of #11911)

### CODE REVIEW

**Comment by daylerees** [comment]
That's pretty cool. I've been looking at ways of injecting through other methods like controllers do. Nice work!

The only thing that's a bit odd here for me is the fluent chaining. While I like fluent chaining, the container is one of those classes that hasn't really done it previously. Unless I'm mistaken.

**Comment by hannesvdvreken** [comment]
@daylerees like `$app->when()->needs()->give()`?

Didn't add that myself, it's something that has been there since, I dunno 5.0? https://laravel.com/docs/5.2/container#contextual-binding

In fact, I didn't have to change a thing there to make it work for both class context and method context.

**Comment by daylerees** [comment]
Ah my bad, I haven't used that yet! Thanks.

File: src/Illuminate/Container/Container.php [hunk 1]
```diff
@@ -565,16 +574,35 @@ protected function getCallReflector($callback)
      * @param  \ReflectionParameter  $parameter
      * @param  array  $parameters
      * @param  array  $dependencies
+     * @param  string $context
+     *
```
Comment (GrahamCampbell) [inline src/Illuminate/Container/Container.php]: please don't add this extra newline

**Comment by hannesvdvreken** [comment]
@daylerees No problem. It's very useful.

I use it a lot like this inside service providers:

```php
$this->app->bind('filesystem.pictures', function () {
    return new Filesystem(new Local(storage_path('pictures')));
});

$this->app->when(Downloader::class)->needs(FilesystemInterface::class)->give('filesystem.pictures');
```

In 2 easy steps you decouple what is injected and what ordinal the argument has has in the constructor.

**Comment by hannesvdvreken** [comment]
@GrahamCampbell fixed, btw

**Comment by hannesvdvreken** [comment]
@taylorotwell @GrahamCampbell What do you think?

**Comment by taylorotwell** [comment]
Pretty interesting. Where do you find yourself needing this feature the most?

**Comment by hannesvdvreken** [comment]
I would use it the most in controllers. But there are many other opportunities. I think it's a logical next step in making method injection and contextual binding possible.

**Comment by hannesvdvreken** [comment]
@taylorotwell @grahamcampbell Always available for a chat about this, if you want. Ping me on slack if needed.

**Comment by andrecadete** [comment]
Any prediction about when this can be merged? (or if it will be at all)

The example @hannesvdvreken gave for the when()->needs()->give() fits perfectly to show how this could be usefull for method injection as well (if slightly changed for multiple contexts):

```
class DownloadsController extends Controller
{

    public method getVideo(DownloaderInterface $downloader, DownloadableFileInterface $file) 
   {
       $downloader->add($file);
       $downloader->start();
   }

    public method getPicture(DownloaderInterface $downloader, DownloadableFileInterface $file) {
        $downloader->add($file);
        $downloader->start();
    }
}

```
and in AppServiceProvider / DownloadsServiceProvider / wtv:

```
$this->app->when('DownloadsController@getVideo')->needs(DownloaderInterface::class)->give(function () {
    return new FileStreamer(storage_path('video'));
});

$this->app->when('DownloadsController@getVideo')->needs(DownloadableFileInterface::class)->give(function () {
    return new DownloadableVideo($this->app('request')->input('input_or_get_param_name'));
});

$this->app->when('DownloadsController@getPicture')->needs(DownloaderInterface::class)->give(function () {
    return new FileDownloader(storage_path('pictures'));
});

$this->app->when('DownloadsController@getPicture')->needs(DownloadableFileInterface::class)->give(function () {
    return new DownloadablePicture($this->app('request')->input('input_or_get_param_name'));
});


```

I'm against magic methods, but I think it could be useful to have this answered for the magic method fans:

Since those 2 methods do exactly the same but simply receive different implementations of the type hinted interfaces, will this work if I use a magic method instead of those 2 copy-paste methods?

**Comment by taylorotwell** [comment]
I guess my counterpoint would be to break that up into two controllers... VideoDownloadController and PictureDownloadController and use constructor injection. Thoughts?

**Comment by hannesvdvreken** [comment]
What about a controller which allows you to download different CSV reports. If you type hint a report with an interface like `Export::getCsv()`. Then you'd have one ExportController with methods `users(Export $export)`, `orders(Export $export)`, ...

**Comment by taylorotwell** [comment]
It seems like you could still just have multiple controllers? Sorry, not trying to be argumentative, but since it's a pretty large amount of complexity to take on into the container I just want to vet it out more.

**Comment by hannesvdvreken** [comment]
No problem. Asked Twitter to help finding use cases. I'm also OK with diffing against master branch for next minor release.

**Comment by andrecadete** [comment]
My only argument @taylorotwell would be that I'd find very logical to have any action under domain.tld/download/ to be in the DownloadControler. But that's not worth much, since it depends on how your team / company works my opinion is probably biased.

**Comment by taylorotwell** [comment]
I'm going to hold off on this because I can't conceive of a situation where you couldn't just break your controller into two smaller controllers to solve this problem without making the container more complex.

**Comment by hannesvdvreken** [comment]
:disappointed:
@daylerees said
> I've been looking at ways of injecting through other methods like controllers do

I bet he's not the only one interested in injecting stuff in other methods throughout his application.

**Comment by andrecadete** [comment]
@taylorotwell I don't disagree at all with your pov that it **CAN **be solved by splitting controllers.. The thing is: The way you organize your Controllers is kind of _tied_ to the business logic, and IMHO I don't think it's healthy for any framework to enforce anything that can be tied to business logic. One will split the Controllers as he wants to and the way you can or cannot inject is tied to the framework and not the Controllers organization. But again, this is an opinion, it's not my framework after all :)

**Comment by andrecadete** [comment]
@taylorotwell you wanted a use case, tell me if it makes any sense at all to split this Controller:

`// Controls actions within domain.tld/foo/
class FooController {

    // lists all Foo (/foo/)
    public function getAll() 
    {
    }

    // shows one Foo (/foo/{id})
    public function getOne() 
    {
    }
    // shows a Foo creation form (/foo/create/)
    public function getCreate(AlerterInterface $alerter) 
    {
    }
    // creates a new Foo (/foo/create/)
    public function postCreate(AlerterInterface $alerter) 
    {
        // I need a SMSAlerter here, if the user can't create a Foo this is super serious for our business,    we need to wake up the SysAd or whatever.
    }
    // deletes existing Foo (/foo/delete/)
    public function postDelete(AlerterInterface $alerter) {
        // I only need a EmailAlerter here, if the user can't delete a Foo this won't affect our sleep. We deal with it tomorrow.
    }

}`

**Pull request closed without merging**